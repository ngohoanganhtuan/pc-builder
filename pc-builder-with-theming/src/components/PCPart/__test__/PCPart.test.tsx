import { fireEvent, render } from '@testing-library/react';
import * as renderer from 'react-test-renderer';

// Components
import { PCPart } from '@components';

// Enums
import { PC_PART_VARIANTS } from '@enums';
import { BrowserRouter } from 'react-router-dom';

describe('Test [PCPart] component', () => {
  test('Selected variant of Component [PCPart] should render correctly', () => {
    const component = renderer
      .create(
        <BrowserRouter>
          <PCPart
            name="PC Part Name"
            price={1000000}
            thumbnail="PC Part Thumbnail"
            variant={PC_PART_VARIANTS.SELECTED}
            quantity={500}
          />
        </BrowserRouter>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  test('Selected variant of Component [PCPart] calls function delete when user click delete button', () => {
    const handleDelete = jest.fn();
    const { getByTestId } = render(
      <BrowserRouter>
        <PCPart
          name="PC Part Name"
          price={1000000}
          thumbnail="PC Part Thumbnail"
          variant={PC_PART_VARIANTS.SELECTED}
          quantity={500}
          handleDelete={handleDelete}
        />
      </BrowserRouter>
    );
    fireEvent.click(getByTestId('pc-part-remove-button'));
    expect(handleDelete).toHaveBeenCalled();
  });

  test('Selected variant of Component [PCPart] calls function increase when user click increase button', () => {
    const handleIncreaseQuantity = jest.fn();
    const { getByTestId } = render(
      <BrowserRouter>
        <PCPart
          name="PC Part Name"
          price={1000000}
          thumbnail="PC Part Thumbnail"
          variant={PC_PART_VARIANTS.SELECTED}
          quantity={500}
          handleIncreaseQuantity={handleIncreaseQuantity}
        />
      </BrowserRouter>
    );
    fireEvent.click(getByTestId('increase-quantity-button'));
    expect(handleIncreaseQuantity).toHaveBeenCalled();
  });

  test('Selected variant of Component [PCPart] calls function decrease when user click decrease button', () => {
    const handleDecreaseQuantity = jest.fn();
    const { getByTestId } = render(
      <BrowserRouter>
        <PCPart
          name="PC Part Name"
          price={1000000}
          thumbnail="PC Part Thumbnail"
          variant={PC_PART_VARIANTS.SELECTED}
          quantity={500}
          handleDecreaseQuantity={handleDecreaseQuantity}
        />
      </BrowserRouter>
    );
    fireEvent.click(getByTestId('decrease-quantity-button'));
    expect(handleDecreaseQuantity).toHaveBeenCalled();
  });

  test('Unselected variant of Component [PCPart] should render correctly', () => {
    const component = renderer
      .create(
        <BrowserRouter>
          <PCPart
            name="PC Part Name"
            price={1000000}
            thumbnail="PC Part Thumbnail"
            variant={PC_PART_VARIANTS.UNSELECTED}
          />
        </BrowserRouter>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  test('Unselected variant of Component [PCPart] call function select when user click select button', () => {
    const handleSelect = jest.fn();
    const { getByTestId } = render(
      <BrowserRouter>
        <PCPart
          name="PC Part Name"
          price={1000000}
          thumbnail="PC Part Thumbnail"
          variant={PC_PART_VARIANTS.UNSELECTED}
          handleSelect={handleSelect}
        />
      </BrowserRouter>
    );
    fireEvent.click(getByTestId('pc-part-select-button'));
    expect(handleSelect).toHaveBeenCalled();
  });
});
