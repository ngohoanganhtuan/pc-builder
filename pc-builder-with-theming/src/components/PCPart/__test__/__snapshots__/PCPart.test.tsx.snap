// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test [PCPart] component Selected variant of Component [PCPart] should render correctly 1`] = `
<div
  className="shadow d-grid pc-part"
>
  <div
    className="p-1"
  >
    <img
      className="border rounded-circle pc-part-thumbnail"
      data-testid="pc-part-thumbnail"
      height="100%"
      src="PC Part Thumbnail"
      width="100%"
    />
  </div>
  <div
    className="p-3 d-flex flex-column align-items-start justify-content-between"
  >
    <a
      className="name card-title fs-6 text-dark fw-bold"
      href="/detail/undefined"
      onClick={[Function]}
    >
      PC Part Name
    </a>
    <div
      className="quantity-input d-flex input-group-sm shadow rounded-pill overflow-hidden quantity-input-sm"
    >
      <button
        className="btn bg-white shadow-none"
        data-testid="decrease-quantity-button"
        onClick={[Function]}
      >
        <i
          className="bi bi-dash"
        />
      </button>
      <div
        className="number bg-white d-flex align-items-center w-100 justify-content-center"
      >
        500
      </div>
      <button
        className="btn bg-white shadow-none"
        data-testid="increase-quantity-button"
        onClick={[Function]}
      >
        <i
          className="bi bi-plus"
        />
      </button>
    </div>
  </div>
  <div
    className="p-3 d-flex flex-column align-items-end justify-content-between"
  >
    <p
      className="fs-6 price"
    >
      1.000.000 ₫
    </p>
    <button
      className="btn btn-sm btn-danger shadow"
      data-testid="pc-part-remove-button"
      onClick={[Function]}
    >
      Remove
    </button>
  </div>
</div>
`;

exports[`Test [PCPart] component Unselected variant of Component [PCPart] should render correctly 1`] = `
<div
  className="shadow d-grid pc-part"
>
  <div
    className="p-1"
  >
    <img
      className="border rounded-circle pc-part-thumbnail"
      data-testid="pc-part-thumbnail"
      height="100%"
      src="PC Part Thumbnail"
      width="100%"
    />
  </div>
  <div
    className="p-3 d-flex flex-column align-items-start justify-content-between"
  >
    <a
      className="name card-title fs-6 text-dark fw-bold"
      href="/detail/undefined"
      onClick={[Function]}
    >
      PC Part Name
    </a>
  </div>
  <div
    className="p-3 d-flex flex-column align-items-end justify-content-between"
  >
    <p
      className="fs-6 price"
    >
      1.000.000 ₫
    </p>
    <button
      className="btn btn-sm btn-success shadow"
      data-testid="pc-part-select-button"
    >
      Select
    </button>
  </div>
</div>
`;
